chainerrl/agents/trpo.py

```diff
@@ -0,0 +1,572 @@
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+from __future__ import absolute_import
+from builtins import *  # NOQA
+from future import standard_library
+standard_library.install_aliases()
+
+import collections
+from distutils.version import StrictVersion
+import itertools
+from logging import getLogger
+
+import chainer
+import chainer.functions as F
+import numpy as np
+
+import chainerrl
+from chainerrl import agent
+from chainerrl.misc.batch_states import batch_states
+
+
+_is_double_backprop_supported = (
+    StrictVersion(chainer.__version__) >= StrictVersion('3.0.0'))
+
+
+def _get_ordered_params(link):
+    """Get a list of parameters sorted by parameter names."""
+    name_param_pairs = list(link.namedparams())
+    ordered_name_param_pairs = sorted(name_param_pairs, key=lambda x: x[0])
+    return [x[1] for x in ordered_name_param_pairs]
+
+
+def _chainer_grad_with_zero(outputs, inputs, *args, **kwargs):
+    """Same as chainer.grad except that None grad will be replaced with zeros.
+
+    Gradients from chainer.grad may contain None. This function replaces None
+    with a zero-filled variable with the same shape as the corresponding input
+    variable.
+    """
+    default_grads = chainer.grad(outputs, inputs, *args, **kwargs)
+    xp = chainer.cuda.get_array_module(inputs[0])
+    return [grad if grad is not None
+            else chainer.Variable(xp.zeros_like(param.data))
+            for grad, param in zip(default_grads, inputs)]
+
+
+def _flatten_and_concat_variables(vs):
+    """Flatten and concat variables to make a single flat vector variable."""
+    return F.concat([F.flatten(v) for v in vs], axis=0)
+
+
+def _as_ndarray(x):
+    """chainer.Variable or ndarray -> ndarray."""
+    if isinstance(x, chainer.Variable):
+        return x.data
+    else:
+        return x
+
+
+def _flatten_and_concat_ndarrays(vs):
+    """Flatten and concat variables to make a single flat vector ndarray."""
+    xp = chainer.cuda.get_array_module(vs[0])
+    return xp.concatenate([_as_ndarray(v).ravel() for v in vs], axis=0)
+
+
+def _split_and_reshape_to_ndarrays(flat_v, sizes, shapes):
+    """Split and reshape a single flat vector to make a list of ndarrays."""
+    xp = chainer.cuda.get_array_module(flat_v)
+    sections = np.cumsum(sizes)
+    vs = xp.split(flat_v, sections)
+    return [v.reshape(shape) for v, shape in zip(vs, shapes)]
+
+
+def _replace_params_data(params, new_params_data):
+    """Replace data of params with new data."""
+    for param, new_param_data in zip(params, new_params_data):
+        assert param.shape == new_param_data.shape
+        param.data[:] = new_param_data
+
+
+def _hessian_vector_product(flat_grads, params, vec):
+    """Compute hessian vector product efficiently by backprop."""
+    grads = _chainer_grad_with_zero([F.sum(flat_grads * vec)], params)
```

Comment:
Right. We cannot assume that for general hessian vector product, but it's true TRPO doesn't work with unused parameters because of CG. I think `chainer.grad` and raising an informative error when there's None would be better. I'll fix it.